edk2toolext/rust_environment.py

```diff
@@ -0,0 +1,442 @@
+    """Returns any tools and their required versions from the workspace toolchain file.
+
+    Returns:
+        dict[str,str]: dict where the key is the tool name and the
+        value is the version
+    """
+    tool_ver = {}
+    try:
+        with open(WORKSPACE_TOOLCHAIN_FILE, "r") as toml_file:
+            content = toml_file.read()
+            match = re.search(r"\[tool\]\n((?:.+\s*=\s*.+\n?)*)", content)
+            if match:
+                for line in match.group(1).splitlines():
+                    (tool, version) = line.split("=", maxsplit=1)
+                    tool_ver[tool.strip()] = version.strip(" \"'")
+        return tool_ver
+    except FileNotFoundError:
+        # If a file is not found. Do not check any further.
+        return tool_ver
+
+
+def _verify_rust_src_component_is_installed() -> bool:
+    """Verifies the rust-src component is installed.
+
+    Returns:
+        bool: True if the rust-src component is installed for the default
+        toolchain or the status could not be determined, otherwise, False.
+    """
+    toolchain_version = get_workspace_toolchain_version()
+    if toolchain_version.error or not toolchain_version:
+        # If the file is not in an expected format, let that be handled
+        # elsewhere and do not look further.
+        return True
+
+    toolchain_version = toolchain_version.toolchain
+
+    rustc_output = StringIO()
+    ret = RunCmd(
+        "rustc",
+        "--version --verbose",
+        outstream=rustc_output,
+        logging_level=logging.DEBUG,
+    )
+    if ret != 0:
+        # rustc installation is checked elsewhere. Exit here on failure.
+        return True
+
+    for line in rustc_output.getvalue().splitlines():
+        start_index = line.lower().strip().find("host: ")
+        if start_index != -1:
+            target_triple = line[start_index + len("host: ") :]
+            break
+    else:
+        logging.error("Failed to get host target triple information.")
+        return False
+
+    rustup_output = StringIO()
+    ret = RunCmd(
+        "rustup",
+        f"component list --toolchain {toolchain_version}",
+        outstream=rustup_output,
+        logging_level=logging.DEBUG,
+    )
+    if ret != 0:
+        # rustup installation and the toolchain are checked elsewhere.
+        # Exit here on failure.
+        return True
+
+    for component in rustup_output.getvalue().splitlines():
+        if "rust-src (installed)" in component:
+            return True
+
+    logging.error(
+        "The Rust toolchain is installed but the rust-src component "
+        "needs to be installed:\n\n"
+        f"  rustup component add --toolchain {toolchain_version}-"
+        f"{target_triple} rust-src"
+    )
+
+    return False
+
+
+def verify_workspace_rust_toolchain_is_installed() -> RustToolChainInfo:
+    """Verifies the rust toolchain used in the workspace is available.
+
+    Note: This function does not use the toml library to parse the toml
```

Comment:
The below suggestion will add a "pretty" note, see an example [here](http://127.0.0.1:8000/api/edk2_invocable/#edk2toolext.edk2_invocable.Edk2Invocable.GetWorkspaceRoot)
```
!!! note
    This function ...
```